pkg/api/controller.go

```diff
@@ -74,2864 +61,2535 @@ type Controller struct {
-	ctx = context.WithValue(ctx, UserContextKey, user)
-	return ctx, authorize(ctx, c.Auth, user, permissions)
-}
-
-func createPaginator(nextToken string, amountResults int) *models.Pagination {
-	return &models.Pagination{
-		HasMore:    swag.Bool(nextToken != ""),
-		MaxPerPage: swag.Int64(MaxResultsPerPage),
-		NextOffset: nextToken,
-		Results:    swag.Int64(int64(amountResults)),
-	}
-}
-
-func pageAmount(i *int64) int {
-	inti := int(swag.Int64Value(i))
-	if inti > MaxResultsPerPage {
-		return MaxResultsPerPage
-	}
-	if inti <= 0 {
-		return DefaultResultsPerPage
-	}
-	return inti
-}
-
-func (c *Controller) GetHealthCheckHandler() hcop.HealthCheckHandler {
-	return hcop.HealthCheckHandlerFunc(func(params hcop.HealthCheckParams) middleware.Responder {
-		return hcop.NewHealthCheckNoContent()
+func (c *Controller) ListGroups(w http.ResponseWriter, r *http.Request, params ListGroupsParams) {
+	if !c.authorize(w, r, []permissions.Permission{
+		{
+			Action:   permissions.ListGroupsAction,
+			Resource: permissions.All,
+		},
+	}) {
+		return
+	}
+
+	ctx := r.Context()
+	c.LogAction(ctx, "list_groups")
+	groups, paginator, err := c.Auth.ListGroups(ctx, &model.PaginationParams{
+		After:  paginationAfter(params.After),
+		Amount: paginationAmount(params.Amount),
 	})
-}
+	if handleAPIError(w, err) {
+		return
+	}
 
-func (c *Controller) SetupLakeFSHandler() setupop.SetupLakeFSHandler {
-	return setupop.SetupLakeFSHandlerFunc(func(setupReq setupop.SetupLakeFSParams) middleware.Responder {
-		if len(*setupReq.User.Username) == 0 {
-			return setupop.NewSetupLakeFSBadRequest().
-				WithPayload(&models.Error{
-					Message: "empty display name",
-				})
-		}
+	response := GroupList{
+		Results: make([]Group, 0, len(groups)),
+		Pagination: Pagination{
+			NextOffset: paginator.NextPageToken,
+			Results:    paginator.Amount,
+		},
+	}
 
-		// check if previous setup completed
-		if ts, _ := c.MetadataManager.SetupTimestamp(setupReq.HTTPRequest.Context()); !ts.IsZero() {
-			return setupop.NewSetupLakeFSConflict().
-				WithPayload(&models.Error{
-					Message: "lakeFS already initialized",
-				})
-		}
+	for _, g := range groups {
+		response.Results = append(response.Results, Group{
+			Id:           g.DisplayName,
+			CreationDate: g.CreatedAt.Unix(),
+		})
+	}
+	writeResponse(w, http.StatusOK, response)
+}
 
-		// migrate the database if needed
-		ctx := setupReq.HTTPRequest.Context()
-		err := c.Migrator.Migrate(ctx)
-		if err != nil {
-			return setupop.NewSetupLakeFSDefault(http.StatusInternalServerError).
-				WithPayload(&models.Error{
-					Message: err.Error(),
-				})
-		}
+func (c *Controller) CreateGroup(w http.ResponseWriter, r *http.Request, body CreateGroupJSONRequestBody) {
+	if !c.authorize(w, r, []permissions.Permission{
+		{
+			Action:   permissions.CreateGroupAction,
+			Resource: permissions.GroupArn(body.Id),
+		},
+	}) {
+		return
+	}
+	ctx := r.Context()
+	c.LogAction(ctx, "create_group")
```

Comment:
It's soooooo much better if you could push this repeatable error prune operation to a generic middleware.

Reply:
If possible to do the same without losing the BI, prefer to do it in a different PR.